*******> update.1

Author: David Case
Date: October 17, 2024
Programs: UseMiniconda.cmake

Description: 1) Pin numpy to version 1.26.4; many amber python scripts don't 
                work with numpy version 2
             2) Replace the minconda download with miniforge3

--------------------------------------------------------------------------------

 cmake/UseMiniconda.cmake                | 32 +++++++++++++++++++++-----------
 1 files changed, 32 insertions(+), 16 deletions(-)

diff --git cmake/UseMiniconda.cmake cmake/UseMiniconda.cmake
index a903a06184..1c453edfbe 100644
--- cmake/UseMiniconda.cmake
+++ cmake/UseMiniconda.cmake
@@ -1,5 +1,9 @@
-# Downloads and installs a Miniconda appropriate for your operating system
+# Downloads and installs a Miniforge appropriate for your operating system
 # This script does not work when crosscompiling.
+#  Note: to make the initial changes as limited as possible the phrase
+#   "MINICONDA" is retained for now, but the actual download and
+#   installation comes from Miniforge3.  Only tests on Linux x86_64 so 
+#   far, not (for example) on OSX.
 
 # Send the version variables up one scope level from the caller of this macro
 macro(proxy_python_version)
@@ -34,17 +38,19 @@ function(download_and_use_miniconda)
 	endif()
 	
 	set(MINICONDA_PYTHON ${MINICONDA_PYTHON} PARENT_SCOPE)
+    # Always use the conda-forge channel so we comply with Anaconda ToS
+    set(ENV{CONDA_CHANNELS} conda-forge)
 	
 	file(MAKE_DIRECTORY ${MINICONDA_TEMP_DIR} ${MINICONDA_DOWNLOAD_DIR})
 	
 	# check if we have already downloaded miniconda
 	if(EXISTS ${MINICONDA_STAMP_FILE})
 		proxy_python_version()
-		message(STATUS "Miniconda is installed in the build directory!")
+		message(STATUS "Miniforge3 is installed in the build directory!")
 		return()
 	endif()
 
-    message(STATUS "Downloading Python 3 Miniconda")	    
+    message(STATUS "Downloading Python 3 Miniforge")	    
 
 	# Figure out the OS part of the URL
 	if(TARGET_OSX)
@@ -70,7 +76,7 @@ function(download_and_use_miniconda)
 	# Figure out the bitiness part of the URL
 	if(TARGET_OSX)
 		if("${TARGET_ARCH}" STREQUAL "x86_64")
-			message(STATUS "Using 64 bit miniconda")
+			message(STATUS "Using 64 bit miniforge3")
 			set(CONTINUUM_BITS "x86_64")
 		elseif("${TARGET_ARCH}" MATCHES "arm64.*")
 			message(STATUS "Using arm64 miniconda")
@@ -78,7 +84,7 @@ function(download_and_use_miniconda)
 		endif()
 	else()
 		if("${TARGET_ARCH}" STREQUAL "x86_64")
-			message(STATUS "Using 64 bit miniconda")
+			message(STATUS "Using 64 bit miniforge3")
 			set(CONTINUUM_BITS "x86_64")
 		elseif("${TARGET_ARCH}" STREQUAL "i386")
 			message(STATUS "Using 32 bit miniconda")
@@ -92,22 +98,22 @@ function(download_and_use_miniconda)
 		endif()
 	endif()
 	
-	set(MINICONDA_INSTALLER_FILENAME "Miniconda${PYTHON_MAJOR_RELEASE}-${MINICONDA_VERSION}-${CONTINUUM_SYSTEM_NAME}-${CONTINUUM_BITS}.${INSTALLER_SUFFIX}")
+	set(MINICONDA_INSTALLER_FILENAME "Miniforge3-${CONTINUUM_SYSTEM_NAME}-${CONTINUUM_BITS}.${INSTALLER_SUFFIX}")
 	
 	# location to download the installer to
 	set(MINICONDA_INSTALLER ${MINICONDA_DOWNLOAD_DIR}/${MINICONDA_INSTALLER_FILENAME})
-	set(INSTALLER_URL "http://repo.continuum.io/miniconda/${MINICONDA_INSTALLER_FILENAME}")
+	set(INSTALLER_URL "https://github.com/conda-forge/miniforge/releases/latest/download/${MINICONDA_INSTALLER_FILENAME}")
 	
 	# If we've already downloaded the installer, use it.	
 	if(EXISTS "${MINICONDA_INSTALLER}")
-		message(STATUS "Using cached Miniconda installer at ${MINICONDA_INSTALLER}")
+		message(STATUS "Using cached Miniforge3 installer at ${MINICONDA_INSTALLER}")
 	else()
 		message("Downloading ${INSTALLER_URL} -> ${MINICONDA_INSTALLER}")
 			
 		# Actually download the file
 		download_file_https(${INSTALLER_URL} ${MINICONDA_INSTALLER} TRUE)
 	endif()
-	message("Installing Miniconda Python.")
+	message("Installing Miniforge Python.")
 	
 	# get rid of the install directory, if it exists
 	file(REMOVE_RECURSE ${MINICONDA_INSTALL_DIR})
@@ -152,10 +158,14 @@ function(download_and_use_miniconda)
 	# if the miniconda version has been specified (-DMINICONDA_VERSION=...)
 	# then do not update conda
 	if(MINICONDA_AUTO)
+    
+		# not needed for miniforge:
+		# execute_process(COMMAND ${CONDA} install -y --solver=classic conda-forge::conda-libmamba-solver conda-forge::libmamba conda-forge::libmambapy conda-forge::libarchive)
+		# execute_process(COMMAND ${CONDA} update --all -y)
 		execute_process(COMMAND ${CONDA} update conda -y)
 	endif()
 	execute_process(COMMAND ${MINICONDA_PYTHON} -m pip install pip --upgrade)
-	
+
 	# Prefer non-mkl packages.
 	# This is because if Amber is using MKL, when Python programs run they will try to talk to two
 	# different MKL libraries at the same time: the MKL Miniconda python was linked with, and the MKL
@@ -166,7 +176,7 @@ function(download_and_use_miniconda)
 	execute_process(COMMAND ${CONDA} install -y -c conda-forge f90nml mrcfile pdb2pqr)
 	execute_process(COMMAND ${CONDA} install -y pandas)
 	
-	execute_process(COMMAND ${CONDA} install -y -q conda-build numpy scipy cython=0.29 ipython notebook pytest 
+	execute_process(COMMAND ${CONDA} install -y -q conda-build numpy=1.26.4 scipy cython=0.29 ipython notebook pytest 
 		RESULT_VARIABLE PACKAGE_INSTALL_RETVAL)
 	if(NOT ${PACKAGE_INSTALL_RETVAL} EQUAL 0)
 		message(FATAL_ERROR "Installation of packages failed!  Please fix what's wrong, or disable Miniconda.")
